Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryManager.java

```diff
@@ -154,39 +172,179 @@ else if (library != null) {
+            e.printStackTrace();
+        }
+
+        if (library != null &&
+                (translatorOptionsMatch(library, options))) {
+            translatedLibrary = generateTranslatedLibrary(library);
+            if (translatedLibrary != null) {
+                this.cacheLibrary(translatedLibrary);
+            }
+        }
+
+        return translatedLibrary;
+    }
+
+    private TranslatedLibrary generateTranslatedLibrary(Library library) {
+
+        if (library == null) {
+            return null;
+        }
+        boolean translationSuccess = true;
+        TranslatedLibrary translatedLibrary = new TranslatedLibrary();
+        try {
+            if (library != null) {
+                translatedLibrary.setLibrary(library);
+            }
+            if (library.getIdentifier() != null) {
+                translatedLibrary.setIdentifier(library.getIdentifier());
+            }
+
+            if (library.getUsings() != null && library.getUsings().getDef() != null) {
+                for (UsingDef usingDef : library.getUsings().getDef()) {
+                    translatedLibrary.add(usingDef);
+                }
+            }
+            if (library.getIncludes() != null && library.getIncludes().getDef() != null) {
+                for (IncludeDef includeDef : library.getIncludes().getDef()) {
+                    translatedLibrary.add(includeDef);
+                }
+            }
+            if (library.getCodeSystems() != null && library.getCodeSystems().getDef() != null) {
+                for (CodeSystemDef codeSystemDef : library.getCodeSystems().getDef()) {
+                    translatedLibrary.add(codeSystemDef);
+                }
+            }
+            for (ValueSetDef valueSetDef : library.getValueSets().getDef()) {
+                translatedLibrary.add(valueSetDef);
+            }
+
+            if (library.getCodes() != null && library.getCodes().getDef() != null) {
+                for (CodeDef codeDef : library.getCodes().getDef()) {
+                    translatedLibrary.add(codeDef);
+                }
+            }
+            if (library.getConcepts() != null && library.getConcepts().getDef() != null) {
+                for (ConceptDef conceptDef : library.getConcepts().getDef()) {
+                    translatedLibrary.add(conceptDef);
+                }
+            }
+            if (library.getParameters() != null && library.getParameters().getDef() != null) {
+                for (ParameterDef parameterDef : library.getParameters().getDef()) {
+                    translatedLibrary.add(parameterDef);
+                }
+            }
+            if (library.getStatements() != null && library.getStatements().getDef() != null) {
+                for (ExpressionDef expressionDef : library.getStatements().getDef()) {
+                    translatedLibrary.add(expressionDef);
+                }
+            }
+        } catch (Exception e) {
+            translationSuccess = false;
```

Comment:
This needs a check to make sure that the resultType is set for each node. This is probably best done as a visitor implementation that would actually have utility elsewhere as well, so I suggest implementing an ElmTypeInferencingVisitor and then using that. I don't think that should hold up this PR, but we at least need a check here to make sure that if the ELM doesn't have resultType for at least each expressionDef that it can't be used as a source, and translationSuccess should be false.